sway-core/src/semantic_analysis/ast_node/declaration/auto_impl/debug.rs

```diff
@@ -0,0 +1,147 @@
+use sway_error::handler::Handler;
+use sway_types::{BaseIdent, Named, Spanned};
+
+use crate::{
+    decl_engine::DeclEngineGet,
+    language::ty::{self, TyAstNode, TyDecl, TyEnumDecl, TyStructDecl},
+    Engines, TypeParameter,
+};
+
+use super::abi_encoding::AbiEncodingAutoImplInfo;
+
+pub type DebugAutoImplContext<'a, 'b> = super::AutoImplContext<'a, 'b, AbiEncodingAutoImplInfo>;
+
+impl<'a, 'b> DebugAutoImplContext<'a, 'b>
+where
+    'a: 'b,
+{
+    pub fn generate_debug_impl(
+        &mut self,
+        engines: &Engines,
+        decl: &ty::TyDecl,
+    ) -> Option<TyAstNode> {
+        match decl {
+            TyDecl::StructDecl(_) => self.debug_auto_impl_struct(engines, decl).unwrap_or(None),
+            TyDecl::EnumDecl(_) => self.debug_auto_impl_enum(engines, decl).unwrap_or(None),
+            _ => None,
+        }
+    }
+
+    // Auto implements Debug for structs and returns their `AstNode`s.
+    fn debug_auto_impl_struct(
+        &mut self,
+        engines: &Engines,
+        decl: &TyDecl,
+    ) -> Option<Option<TyAstNode>> {
```

Comment:
It doesn't look like the first Option is of any use here, is it just a usage convenience?

Reply:
You are right. Done.